### PR TITLE
CPU Batch Norm Op

### PR DESCRIPTION
**Pull request by jacquesqiao**
## Background
project: https://github.com/PaddlePaddle/Paddle/issues/4531
fix: https://github.com/PaddlePaddle/Paddle/issues/4906

## Progress
- [x] CPU forward
- [x] CPU gradient
- [x] Support NCHW && NHWC
- [x] Python Unit Test(forward && gradient)

### CODE REVIEW

File: paddle/operators/batch_norm_op.cc [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+    const std::string tensor_format_str =
+        ctx.Attr<std::string>("tensor_format");
+    const TensorFormat tensor_format = StringToTensorFormat(tensor_format_str);
+
+    const auto *x = ctx.Input<Tensor>("X");
+    const auto &x_dims = x->dims();
+
+    PADDLE_ENFORCE(x_dims.size() >= 3 && x_dims.size() <= 5,
+                   "The Input dim size should be between 3 and 5");
+    const int N = x_dims[0];
```
Comment (zchen0211) [inline paddle/operators/batch_norm_op.cc]: Do we need to enforce Google coding style?

File: paddle/operators/batch_norm_op.cc [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+    if (!is_test) {
+      // saved_xx is use just in this batch of data
+      EigenVectorArrayMap<T> saved_mean_e(
+          saved_mean->mutable_data<T>(ctx.GetPlace()), C);
+      EigenVectorArrayMap<T> saved_variance_e(
+          saved_variance->mutable_data<T>(ctx.GetPlace()), C);
+      saved_mean_e.setZero();
+      saved_variance_e.setZero();
+
+      switch (tensor_format) {
```
Comment (zchen0211) [inline paddle/operators/batch_norm_op.cc]: A complete implementation, which is very nice. Do we only need to do NCHW or NHWC? We can discuss.

Comment (jacquesqiao) [reply paddle/operators/batch_norm_op.cc]: I find our old code support both format,  and Tensorflow/Caffe2 also support two data format, so I think maybe we need them all.

**Comment by zhouxiao-coder** [comment]
Do we need to support 2D input data? For example, if the previous layer is a fully connected layer.

File: paddle/operators/batch_norm_op.cc [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+    const int W =
+        x_dims.size() > 3
+            ? (tensor_format == TensorFormat::NCHW ? x_dims[3] : x_dims[2])
+            : 1;
+    const int D =
+        x_dims.size() > 4
+            ? (tensor_format == TensorFormat::NCHW ? x_dims[4] : x_dims[3])
+            : 1;
+
+    const int sample_size = H * W * D;
```
Comment (qingqing01) [inline paddle/operators/batch_norm_op.cc]: 输入是5D: 即 `NCDHW` 或 `NDHWC`:
可以去掉line 136 - line 144复杂的获取方法，line 146改成：

```cpp
const int frame_size =  x->numel() / N / C;
```

File: paddle/operators/batch_norm_op.cc (hunk 2 above)
Comment (qingqing01) [reply paddle/operators/batch_norm_op.cc]: Our old code only support `NCHW` (for 2D Conv) or `NCDHW` (for 3D Conv), not support `NHWC` and `NDHWC`.  And all the convolution operators in new framework only support NCHW` (for 2D Conv) or `NCDHW` too.

File: paddle/operators/batch_norm_op.cc [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+      EigenVectorArrayMap<T> saved_mean_e(
+          saved_mean->mutable_data<T>(ctx.GetPlace()), C);
+      EigenVectorArrayMap<T> saved_variance_e(
+          saved_variance->mutable_data<T>(ctx.GetPlace()), C);
+      saved_mean_e.setZero();
+      saved_variance_e.setZero();
+
+      switch (tensor_format) {
+        case TensorFormat::NCHW: {
+          ConstEigenArrayMap<T> X_arr(x->data<T>(), sample_size, N * C);
```
Comment (qingqing01) [inline paddle/operators/batch_norm_op.cc]: X_arr -> x_arr :  https://google.github.io/styleguide/cppguide.html#Variable_Names

File: paddle/operators/batch_norm_op.cc [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+          for (int nc = 0; nc < N * C; ++nc) {
+            saved_mean_e(nc % C) += X_arr.col(nc).sum();
+          }
+          saved_mean_e /= N * sample_size;
+          for (int nc = 0; nc < N * C; ++nc) {
+            saved_variance_e(nc % C) +=
+                (X_arr.col(nc) - saved_variance_e(nc % C))
```
Comment (qingqing01) [inline paddle/operators/batch_norm_op.cc]: saved_variance_e ->  saved_mean_e?
```cpp
saved_variance_e(nc % C) +=
    (X_arr.col(nc) - saved_mean_e(nc % C))
```

File: paddle/operators/batch_norm_op.cc [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+    const int W =
+        x_dims.size() > 3
+            ? (tensor_format == TensorFormat::NCHW ? x_dims[3] : x_dims[2])
+            : 1;
+    const int D =
+        x_dims.size() > 4
+            ? (tensor_format == TensorFormat::NCHW ? x_dims[4] : x_dims[3])
+            : 1;
+
+    const int sample_size = H * W * D;
```
Comment (qingqing01) [inline paddle/operators/batch_norm_op.cc]: same as above.

File: paddle/operators/batch_norm_op.cc [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+
+    ConstEigenVectorArrayMap<T> scale_arr(scale->data<T>(), C);
+    ConstEigenVectorArrayMap<T> mean_arr(saved_mean->data<T>(), C);
+    ConstEigenVectorArrayMap<T> inv_var_arr(saved_inv_variance->data<T>(), C);
+
+    // init output
+    auto *dX = ctx.Output<Tensor>(framework::GradVarName("X"));
+    auto *dScale = ctx.Output<Tensor>(framework::GradVarName("Scale"));
+    auto *dBias = ctx.Output<Tensor>(framework::GradVarName("Bias"));
```
Comment (qingqing01) [inline paddle/operators/batch_norm_op.cc]: > The names of variables (including function parameters) and data members are all lowercase

https://google.github.io/styleguide/cppguide.html#Variable_Names

File: paddle/operators/batch_norm_op.cc [hunk 8]
```diff
@@ -0,0 +1,424 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
```
Comment (wangkuiyi) [inline paddle/operators/batch_norm_op.cc]: Reserve => Reserved.

File: paddle/operators/batch_norm_op.h [hunk 9]
```diff
@@ -0,0 +1,62 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
```
Comment (wangkuiyi) [inline paddle/operators/batch_norm_op.h]: Reserve => Reserved

File: paddle/operators/batch_norm_op.cc (hunk 2 above)
Comment (jacquesqiao) [reply paddle/operators/batch_norm_op.cc]: 👌

File: paddle/operators/batch_norm_op.cc (hunk 8 above)
Comment (jacquesqiao) [reply paddle/operators/batch_norm_op.cc]: all the code is wrong now, I will update them in another PR

File: paddle/operators/batch_norm_op.cc (hunk 1 above)
Comment (jacquesqiao) [reply paddle/operators/batch_norm_op.cc]: done

File: paddle/operators/batch_norm_op.h [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+template <typename Place, typename T>
+class BatchNormKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override;
+};
+
+enum TensorFormat {
+  NHWC = 0,
+  NCHW = 1,
+};
```
[the remark this answers is not in the document]
Comment (jacquesqiao) [reply paddle/operators/batch_norm_op.h]: cudnn_helper cannot be included in a cc file, I think maybe we need to move it to another place
